diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt

```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
     private fun isSingleLineStatement(node: ASTNode): Boolean =
             !node.text.contains("\n")
 
-    private fun detectStringConcatenation(node: ASTNode) {
-        node.findAllNodesWithSpecificType(BINARY_EXPRESSION).find { detectStringConcatenationInExpression(it, node) }
+    /**
+     * This method works only with top-level binary expressions. It should be checked before the call.
+     */
+    private fun detectStringConcatenation(topLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = topLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { isDetectStringConcatenationInExpression(it) }
+
+        val bugDetected = nodeWithBug != null
+
+        if (bugDetected) {
+            STRING_CONCATENATION.warnAndFix(
+                configRules, emitWarn,
+                this.isFixMode, topLevelBinaryExpr.text, nodeWithBug!!.startOffset, nodeWithBug
+            ) {
+                circle(topLevelBinaryExpr.treeParent)
+            }
+        }
     }
 
-    @Suppress("FUNCTION_BOOLEAN_PREFIX")
-    private fun detectStringConcatenationInExpression(node: ASTNode, parentNode: ASTNode): Boolean {
-        assert(node.elementType == BINARY_EXPRESSION)
-        val firstChild = node.firstChildNode
-        return if (isPlusBinaryExpression(node) && firstChild.elementType == STRING_TEMPLATE) {
-            STRING_CONCATENATION.warn(configRules, emitWarn, this.isFixMode, parentNode.text, firstChild.startOffset, firstChild)
-            true
-        } else {
-            false
+    private fun circle(parentTopLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = parentTopLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { isDetectStringConcatenationInExpression(it) }
+
+        val bugDetected = nodeWithBug != null
+        if (bugDetected) {
+            fixBinaryExpressionWithConcatenation(nodeWithBug)
+            circle(parentTopLevelBinaryExpr)
         }
     }
 
-    @Suppress("COMMENT_WHITE_SPACE")
+    /**
+     * We can detect string concatenation by the first (left) operand in binary expression.
+     * If it is of type string - then we found string concatenation.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        assert(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
+        }
+        val firstChild = node.firstChildNode
+        return (isPlusBinaryExpression(node) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        (firstChild.text.contains("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
```

Comment:
```suggestion
                        (firstChild.text.endsWith("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
```
There can be cases when resulting string is transformed into something else